backend/jobs/workers/bigquery/bq_to_ads_offline_click_conversion.py

```diff
@@ -0,0 +1,228 @@
+     'bool',
+     False,
+     False,
+     'Flag determining if each conversion upload response should be logged'),
+  ]
+
+  GLOBAL_SETTINGS = [
+    DEVELOPER_TOKEN,
+    SERVICE_ACCOUNT_FILE,
+    REFRESH_TOKEN,
+    CLIENT_ID,
+    CLIENT_SECRET,
+  ]
+
+  def _validate_params(self) -> None:
+    err_messages = []
+
+    err_messages += self._validate_bq_params()
+    err_messages += self._validate_ads_client_params()
+
+    if err_messages:
+      raise ValueError('The following param validation errors occurred:\n' +
+                       '\n'.join(err_messages))
+
+  def _validate_bq_params(self) -> List[str]:
+    err_messages = []
+
+    if not self._params.get(bq_worker.BQ_PROJECT_ID_PARAM_NAME, None):
+      err_messages.append(
+        '"'+bq_worker.BQ_PROJECT_ID_PARAM_NAME+'" is required.')
+
+    if not self._params.get(bq_worker.BQ_DATASET_NAME_PARAM_NAME, None):
+      err_messages.append(
+        '"'+bq_worker.BQ_DATASET_NAME_PARAM_NAME+'" is required.')
+
+    if not self._params.get(bq_worker.BQ_TABLE_NAME_PARAM_NAME, None):
+      err_messages.append(
+        '"'+bq_worker.BQ_TABLE_NAME_PARAM_NAME+'" is required.')
+
+    return err_messages
+
+  def _validate_ads_client_params(self) -> List[str]:
+    err_messages = []
+    is_service_account = SERVICE_ACCOUNT_FILE in self._params
+
+    if not self._params.get(SERVICE_ACCOUNT_FILE, None) and not self._params.get(REFRESH_TOKEN, None):
+      err_messages.append(
+        'Either "'+SERVICE_ACCOUNT_FILE+'" or "'+REFRESH_TOKEN+'" are required')
+
+    if not self._params.get(CONVERSION_UPLOAD_JSON_TEMPLATE, None):
+      err_messages.append('"'+CONVERSION_UPLOAD_JSON_TEMPLATE+'" is required.')
+
+    if not self._params.get(CONVERSIONS_CUSTOMER_ID, None):
+      err_messages.append('"'+CONVERSIONS_CUSTOMER_ID+'" is required.')
+
+    if not self._params.get(DEVELOPER_TOKEN, None):
+      err_messages.append('"'+DEVELOPER_TOKEN+'" is required.')
+
+    if not is_service_account and not self._params.get(CLIENT_ID, None) :
+      err_messages.append('"'+CLIENT_ID+'" is required.')
+
+    if not is_service_account and not self._params.get(CLIENT_SECRET, None):
+      err_messages.append('"'+CLIENT_SECRET+'" is required.')
+
+    return err_messages
+
+  def _execute(self) -> None:
+    """Begin the processing and upload of offline click conversions."""
+    self.log_info('Validating parameters now.')
+    self._validate_params()
+    super()._execute()
+
+  def _get_sub_worker_name(self) -> str:
+    return AdsOfflineClickPageResultsWorker.__name__
+
+
+class AdsOfflineClickPageResultsWorker(bq_batch_worker.TablePageResultsProcessorWorker):
+  """A page results worker for uploading a chunk of conversion data."""
+
+  def _process_page_results(self, page_data: page_iterator.Page) -> None:
+    ads_client = self._get_ads_client()
+    num_rows = page_data.num_items
+    template = string.Template(self._params['template'])
+
+    conversion_objs = []
```

Comment:
The limit here is 2000 per request. So, we probably want to have a failsafe here to ensure we don't go over that number.

Reply:
https://developers.google.com/google-ads/api/docs/best-practices/quotas#:~:text=Limited%20to%202%2C000%20call%20or%20click%20conversions%20per%20request